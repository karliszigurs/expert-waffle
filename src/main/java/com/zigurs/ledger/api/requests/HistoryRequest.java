package com.zigurs.ledger.api.requests;

import lombok.NonNull;

import java.util.UUID;

public record HistoryRequest(@NonNull UUID account_id,
                             Integer offset,
                             Integer limit) {
}
