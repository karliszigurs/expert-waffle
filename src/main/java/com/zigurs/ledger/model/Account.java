package com.zigurs.ledger.model;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.UuidGenerator;

import java.math.BigInteger;
import java.util.UUID;

@Getter
@Entity
@Table(name = "accounts")
@AllArgsConstructor
@NoArgsConstructor
public class Account {

    @Id
    @UuidGenerator
    private UUID id;

    @ManyToOne
    private Client client;

    private String currency;

    @Setter
    private BigInteger balance;
}
