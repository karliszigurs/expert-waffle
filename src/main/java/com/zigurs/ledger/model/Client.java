package com.zigurs.ledger.model;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.UuidGenerator;

import java.util.UUID;

@Entity
@Table(name = "clients")
@AllArgsConstructor
@NoArgsConstructor
public class Client {

    @Id
    @UuidGenerator
    @Getter
    private UUID id;
}