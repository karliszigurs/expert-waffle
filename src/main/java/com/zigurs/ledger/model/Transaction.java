package com.zigurs.ledger.model;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.Getter;
import org.hibernate.annotations.UuidGenerator;

import java.math.BigInteger;
import java.time.Instant;
import java.util.UUID;

@Getter
@Entity
@Table(name = "transactions")
public class Transaction {

    @Id
    @UuidGenerator
    private UUID id;
    private TransactionStatus status;
    private Instant timestamp;
    @ManyToOne
    private Account sourceAccount;
    private BigInteger sourceAmount;
    private BigInteger sourceBalance;
    @ManyToOne
    private Account destinationAccount;
    private BigInteger destinationAmount;
    private BigInteger destinationBalance;
    private String description;

    public Transaction() {

    }

    public Transaction(UUID id, TransactionStatus status, Instant timestamp, Account sourceAccount, BigInteger sourceAmount, BigInteger sourceBalance, Account destinationAccount, BigInteger destinationAmount, BigInteger destinationBalance, String description) {
        this.id = id;
        this.status = status;
        this.timestamp = timestamp;
        this.sourceAccount = sourceAccount;
        this.sourceAmount = sourceAmount;
        this.sourceBalance = sourceBalance;
        this.destinationAccount = destinationAccount;
        this.destinationAmount = destinationAmount;
        this.destinationBalance = destinationBalance;
        this.description = description;
    }

}
