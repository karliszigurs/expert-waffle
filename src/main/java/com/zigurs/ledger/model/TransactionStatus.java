package com.zigurs.ledger.model;

public enum TransactionStatus {
    COMPLETED,
    FAILED
}
