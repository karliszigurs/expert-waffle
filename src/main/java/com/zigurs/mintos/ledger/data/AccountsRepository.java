package com.zigurs.mintos.ledger.data;

import com.zigurs.mintos.ledger.model.Account;

import jakarta.persistence.LockModeType;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface AccountsRepository extends Repository<Account, UUID> {

    List<Account> findByClientId(UUID client_id);

    Optional<Account> findById(UUID uuid);

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("select a from Account a where a.id = :id")
    Optional<Account> findByIdForUpdate(UUID id);

    Account save(Account account);
}
