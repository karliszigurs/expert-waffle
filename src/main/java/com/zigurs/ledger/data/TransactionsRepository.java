package com.zigurs.ledger.data;

import com.zigurs.ledger.model.Transaction;

import org.springframework.data.domain.Pageable;
import org.springframework.data.repository.Repository;

import java.util.List;
import java.util.UUID;

public interface TransactionsRepository extends Repository<Transaction, UUID> {

    Transaction save(Transaction transaction);

    List<Transaction> findAllBySourceAccountIdOrDestinationAccountIdOrderByTimestampDesc(UUID sourceAccountId, UUID destinationAccountId, Pageable pageable);
}
