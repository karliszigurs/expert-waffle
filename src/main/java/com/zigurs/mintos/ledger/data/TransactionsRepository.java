package com.zigurs.mintos.ledger.data;

import com.zigurs.mintos.ledger.model.Transaction;

import org.springframework.data.domain.Pageable;
import org.springframework.data.repository.Repository;

import java.util.List;
import java.util.UUID;

public interface TransactionsRepository extends Repository<Transaction, UUID> {

    Transaction save(Transaction transaction);

    List<Transaction> findAllBySourceAccountIdOrDestinationAccountIdOrderByTimestampDesc(UUID sourceAccountId, UUID destinationAccountId, Pageable pageable);
}
