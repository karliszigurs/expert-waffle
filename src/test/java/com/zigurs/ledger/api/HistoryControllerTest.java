package com.zigurs.ledger.api;

import com.zigurs.ledger.api.requests.HistoryRequest;
import com.zigurs.ledger.data.AccountsRepository;
import com.zigurs.ledger.data.TransactionsRepository;
import com.zigurs.ledger.model.Account;
import com.zigurs.ledger.model.Client;
import com.zigurs.ledger.model.Transaction;
import com.zigurs.ledger.model.TransactionStatus;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.data.domain.Pageable;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import testutils.Snapshot;

import java.math.BigInteger;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static io.github.jsonSnapshot.SnapshotMatcher.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static testutils.Utils.asJsonString;

@WebMvcTest(HistoryController.class)
class HistoryControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private AccountsRepository accountsRepository;

    @MockBean
    private TransactionsRepository transactionsRepository;

    @BeforeAll
    public static void beforeAll() {
        start(Snapshot::asJsonString);
    }

    @AfterAll
    public static void afterAll() {
        validateSnapshots();
    }

    @Test
    void getHistoryHappyPath() throws Exception {
        Client client = new Client(UUID.randomUUID());
        Account account = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        Account differentSourceAccount = new Account(
                UUID.fromString("8c6724de-0879-4392-ba25-cf8dcc79ca36"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(any(UUID.class))).thenReturn(
                Optional.of(account)
        );

        Instant timestamp = Instant.ofEpochSecond(1715933225L);

        when(
                transactionsRepository.findAllBySourceAccountIdOrDestinationAccountIdOrderByTimestampDesc(any(UUID.class), any(UUID.class), any(Pageable.class))
        ).thenReturn(
                List.of(
                        new Transaction(
                                UUID.fromString("424dfead-4625-4dd4-a025-642542eebfdf"),
                                TransactionStatus.COMPLETED,
                                timestamp,
                                account,
                                BigInteger.valueOf(100),
                                BigInteger.valueOf(200),
                                differentSourceAccount,
                                BigInteger.valueOf(300),
                                BigInteger.valueOf(400),
                                "tx one"
                        ),
                        new Transaction(
                                UUID.fromString("ed713c8c-46d1-43b2-9474-63e702983958"),
                                TransactionStatus.COMPLETED,
                                timestamp,
                                differentSourceAccount,
                                BigInteger.valueOf(100),
                                BigInteger.valueOf(200),
                                account,
                                BigInteger.valueOf(300),
                                BigInteger.valueOf(400),
                                "tx two"
                        ),
                        new Transaction(
                                UUID.fromString("5a63b4db-2e30-45bc-a7b7-2dc09f3aaddd"),
                                TransactionStatus.COMPLETED,
                                timestamp,
                                account,
                                BigInteger.valueOf(100),
                                BigInteger.valueOf(200),
                                differentSourceAccount,
                                BigInteger.valueOf(300),
                                BigInteger.valueOf(400),
                                "tx three"
                        ),
                        new Transaction(
                                UUID.fromString("ed49edc4-b1e7-4f89-8bb6-0ef6d917b7da"),
                                TransactionStatus.COMPLETED,
                                timestamp,
                                differentSourceAccount,
                                BigInteger.valueOf(100),
                                BigInteger.valueOf(200),
                                account,
                                BigInteger.valueOf(300),
                                BigInteger.valueOf(400),
                                "tx four"
                        ),
                        new Transaction(
                                UUID.fromString("c6d3cd1f-8fe6-4498-a563-16dbbdaf5bd1"),
                                TransactionStatus.COMPLETED,
                                timestamp,
                                account,
                                BigInteger.valueOf(100),
                                BigInteger.valueOf(200),
                                differentSourceAccount,
                                BigInteger.valueOf(300),
                                BigInteger.valueOf(400),
                                "tx five"
                        )
                )
        );


        HistoryRequest req = new HistoryRequest(UUID.randomUUID(), null, null);

        MvcResult result = mockMvc.perform(
                        post("/history")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is2xxSuccessful())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void getHistoryInvalidAccount() throws Exception {
        when(accountsRepository.findById(any(UUID.class))).thenReturn(
                Optional.empty()
        );

        HistoryRequest req = new HistoryRequest(UUID.randomUUID(), null, null);

        MvcResult result = mockMvc.perform(
                        post("/history")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().isNotFound())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }


    @Test
    void getHistoryInvalidOffset() throws Exception {
        Client client = new Client(UUID.randomUUID());
        Account account = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(any(UUID.class))).thenReturn(
                Optional.of(account)
        );

        // TODO - divide by page ID is breaking check on -1
        HistoryRequest req = new HistoryRequest(UUID.randomUUID(), -100, null);

        MvcResult result = mockMvc.perform(
                        post("/history")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().isBadRequest())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }


    @Test
    void getHistoryInvalidLimit() throws Exception {
        Client client = new Client(UUID.randomUUID());
        Account account = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(any(UUID.class))).thenReturn(
                Optional.of(account)
        );

        // TODO - divide by page ID is breaking check on -1
        HistoryRequest req = new HistoryRequest(UUID.randomUUID(), null, 0);

        MvcResult result = mockMvc.perform(
                        post("/history")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().isBadRequest())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }


    @Test
    void getHistoryValidLimitAndOffset() throws Exception {
        Client client = new Client(UUID.randomUUID());
        Account account = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(any(UUID.class))).thenReturn(
                Optional.of(account)
        );

        when(transactionsRepository.findAllBySourceAccountIdOrDestinationAccountIdOrderByTimestampDesc(any(UUID.class), any(UUID.class), any(Pageable.class))).thenReturn(
                List.of()
        );

        HistoryRequest req = new HistoryRequest(UUID.randomUUID(), 100, 100);

        MvcResult result = mockMvc.perform(
                        post("/history")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().isOk())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }
}