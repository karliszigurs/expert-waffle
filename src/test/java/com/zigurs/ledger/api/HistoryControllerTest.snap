com.zigurs.ledger.api.HistoryControllerTest.getHistoryHappyPath=[
  "{\"account\":{\"account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"currency\":\"EUR\",\"balance\":1},\"offset\":0,\"limit\":10,\"transactions\":[{\"transaction_id\":\"424dfead-4625-4dd4-a025-642542eebfdf\",\"timestamp\":\"2024-05-17T08:07:05Z\",\"type\":\"DEBIT\",\"status\":\"COMPLETED\",\"source_account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"destination_account_id\":\"8c6724de-0879-4392-ba25-cf8dcc79ca36\",\"amount\":100,\"account_balance\":200,\"description\":\"tx one\"},{\"transaction_id\":\"ed713c8c-46d1-43b2-9474-63e702983958\",\"timestamp\":\"2024-05-17T08:07:05Z\",\"type\":\"CREDIT\",\"status\":\"COMPLETED\",\"source_account_id\":\"8c6724de-0879-4392-ba25-cf8dcc79ca36\",\"destination_account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"amount\":300,\"account_balance\":400,\"description\":\"tx two\"},{\"transaction_id\":\"5a63b4db-2e30-45bc-a7b7-2dc09f3aaddd\",\"timestamp\":\"2024-05-17T08:07:05Z\",\"type\":\"DEBIT\",\"status\":\"COMPLETED\",\"source_account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"destination_account_id\":\"8c6724de-0879-4392-ba25-cf8dcc79ca36\",\"amount\":100,\"account_balance\":200,\"description\":\"tx three\"},{\"transaction_id\":\"ed49edc4-b1e7-4f89-8bb6-0ef6d917b7da\",\"timestamp\":\"2024-05-17T08:07:05Z\",\"type\":\"CREDIT\",\"status\":\"COMPLETED\",\"source_account_id\":\"8c6724de-0879-4392-ba25-cf8dcc79ca36\",\"destination_account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"amount\":300,\"account_balance\":400,\"description\":\"tx four\"},{\"transaction_id\":\"c6d3cd1f-8fe6-4498-a563-16dbbdaf5bd1\",\"timestamp\":\"2024-05-17T08:07:05Z\",\"type\":\"DEBIT\",\"status\":\"COMPLETED\",\"source_account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"destination_account_id\":\"8c6724de-0879-4392-ba25-cf8dcc79ca36\",\"amount\":100,\"account_balance\":200,\"description\":\"tx five\"}]}"
]


com.zigurs.ledger.api.HistoryControllerTest.getHistoryInvalidAccount=[
  "{\"error\":\"404 account not found\"}"
]


com.zigurs.ledger.api.HistoryControllerTest.getHistoryInvalidLimit=[
  "{\"error\":\"400 limit must be positive\"}"
]


com.zigurs.ledger.api.HistoryControllerTest.getHistoryInvalidOffset=[
  "{\"error\":\"400 offset must be 0 or positive\"}"
]


com.zigurs.ledger.api.HistoryControllerTest.getHistoryValidLimitAndOffset=[
  "{\"account\":{\"account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"currency\":\"EUR\",\"balance\":1},\"offset\":100,\"limit\":100,\"transactions\":[]}"
]